cpp/src/arrow/buffer-builder.h

```diff
@@ -0,0 +1,205 @@
+// under the License.
+
+#ifndef ARROW_BUFFER_BUILDER_H
+#define ARROW_BUFFER_BUILDER_H
+
+#include <algorithm>
+#include <array>
+#include <cstdint>
+#include <cstring>
+#include <memory>
+#include <string>
+
+#include "arrow/buffer.h"
+#include "arrow/status.h"
+#include "arrow/util/bit-util.h"
+#include "arrow/util/macros.h"
+#include "arrow/util/visibility.h"
+
+namespace arrow {
+
+// ----------------------------------------------------------------------
+// Buffer builder classes
+
+/// \class BufferBuilder
+/// \brief A class for incrementally building a contiguous chunk of in-memory data
+class ARROW_EXPORT BufferBuilder {
+ public:
+  explicit BufferBuilder(MemoryPool* pool ARROW_MEMORY_POOL_DEFAULT)
+      : pool_(pool), data_(NULLPTR), capacity_(0), size_(0) {}
+
+  /// \brief Resize the buffer to the nearest multiple of 64 bytes
+  ///
+  /// \param elements the new capacity of the of the builder. Will be rounded
+  /// up to a multiple of 64 bytes for padding
+  /// \param shrink_to_fit if new capacity is smaller than the existing size,
+  /// reallocate internal buffer. Set to false to avoid reallocations when
+  /// shrinking the builder.
+  /// \return Status
+  Status Resize(const int64_t elements, bool shrink_to_fit = true) {
+    // Resize(0) is a no-op
+    if (elements == 0) {
+      return Status::OK();
+    }
+    int64_t old_capacity = capacity_;
+
+    if (buffer_ == NULLPTR) {
+      ARROW_RETURN_NOT_OK(AllocateResizableBuffer(pool_, elements, &buffer_));
+    } else {
+      ARROW_RETURN_NOT_OK(buffer_->Resize(elements, shrink_to_fit));
+    }
+    capacity_ = buffer_->capacity();
+    data_ = buffer_->mutable_data();
+    if (capacity_ > old_capacity) {
+      memset(data_ + old_capacity, 0, capacity_ - old_capacity);
+    }
+    return Status::OK();
+  }
+
+  /// \brief Ensure that builder can accommodate the additional number of bytes
+  /// without the need to perform allocations
+  ///
+  /// \param size number of additional bytes to make space for
+  /// \return Status
+  Status Reserve(const int64_t size) { return Resize(size_ + size, false); }
+
+  /// \brief Append the given data to the buffer
+  ///
+  /// The buffer is automatically expanded if necessary.
+  Status Append(const void* data, int64_t length) {
+    if (capacity_ < length + size_) {
+      int64_t new_capacity = BitUtil::NextPower2(length + size_);
+      ARROW_RETURN_NOT_OK(Resize(new_capacity));
+    }
+    UnsafeAppend(data, length);
+    return Status::OK();
+  }
+
+  /// \brief Append the given data to the buffer
+  ///
+  /// The buffer is automatically expanded if necessary.
+  template <size_t NBYTES>
+  Status Append(const std::array<uint8_t, NBYTES>& data) {
+    constexpr auto nbytes = static_cast<int64_t>(NBYTES);
+    if (capacity_ < nbytes + size_) {
+      int64_t new_capacity = BitUtil::NextPower2(nbytes + size_);
+      ARROW_RETURN_NOT_OK(Resize(new_capacity));
+    }
+
+    if (nbytes > 0) {
```

Comment:
Sorry for the code churn. What happened here is that I added these guards to avoid calling memset() or memcpy() with a null pointer.